### PR TITLE
change OpenMapSurfer_Roads basemap to terrain

### PR DESCRIPTION
**Pull request by boney-bun**
fix #kartoza/geosafe#531

### CODE REVIEW

**Comment by NyakudyaA** [comment]
@boney-bun does this remove the base maps even at Layer detail page, Because this also affects the thumbnail generation where the basemap is empty

**Review by lucernae — CHANGES_REQUESTED** [review]
Why we don't use OSM as default basemap?
I'm not sure why you add `Terrain` in this page?
Could you use the same logic from Layer details to include set of basemaps taken from the settings file?

**Comment by NyakudyaA** [comment]
> Why we don't use OSM as default basemap?
> I'm not sure why you add `Terrain` in this page?
> Could you use the same logic from Layer details to include set of basemaps taken from the settings file?

Yes @lucernae we just need to have a basemap that we know always works. OSM basemap would be nice to have it as the default and the terrain can still be added as a second option for people who do want to change

**Comment by lucernae** [comment]
> Yes @lucernae we just need to have a basemap that we know always works. OSM basemap would be nice to have it as the default and the terrain can still be added as a second option for people who do want to change

Yeah, argh sorry for not being clear :D. If I read my comment above again, it seems I'm complaining about the `Terrain` map. What I mean is, we should add the `Terrain` in the settings, not in the page. Then take the default basemap from whatever the first basemap in the list from the settings.

In last year's sprint, we already fixed this problem by offloading the basemap to settings, so we made this:

https://github.com/kartoza/geonode/blob/2.8.x-qgis_server/geonode/client/templates/leaflet/layers/layer_leaflet_map.html#L142  (in kartoza/geonode repo)

The layer detail page takes extra basemaps from the settings module.

I suggest that:
1. We take the same approach for analysis related page (because it isn't handled by geonode) in this geosafe repo.
2. Customize the basemaps to everywhere from settings module.
3. Add new settings field to choose default basemaps from a given basemaps defined in settings
4. Refactor layer and map page of GeoNode to choose default basemaps from settings.

The settings that I was talking about:

https://github.com/kartoza/geonode/blob/2.8.x-qgis_server/geonode/settings.py#L1015

It would be nice if we can handle all these

**Comment by boney-bun** [comment]
yes, the default is `osm`. `terrain` is just an additional basemap to replace the removed one.
i have another PR in custom repo.
i thought we will just use the custom one.
i guess i will close the PR in the custom repo and move it to geonode repo.
i also will revise this PR.

**Comment by boney-bun** [comment]
the PR in action:
geosafe#531-geosafe

the basemaps are taken from settings.
the default basemap is always osm.

@lucernae @NyakudyaA could you review again?

File: templates/geosafe/analysis/create_base.html [hunk 1]
```diff
@@ -526,21 +526,22 @@ <h4 class="modal-title" id="myModalLabel">{% trans "Upgrade Your Browser" %}</h4
...
-            };
+            var base_maps = {};
+            {% if LEAFLET_CONFIG and LEAFLET_CONFIG.TILES %}
+                {% for tile in LEAFLET_CONFIG.TILES %}
+
+                    base_maps["{{ tile.0 }}"] = L.tileLayer('{{ tile.1|safe }}', {
+                        attribution: '{{ tile.2|safe }}'
+                });
+                {% endfor %}
+            {% endif %}
```
Comment (lucernae) [inline templates/geosafe/analysis/create_base.html:542]:
```suggestion
            default_basemap = base_maps["{{ LEAFLET_CONFIG.TILES.0 }}"];
            {% endif %}
```

File: templates/geosafe/analysis/create_base.html [hunk 2]
```diff
@@ -526,21 +526,22 @@ <h4 class="modal-title" id="myModalLabel">{% trans "Upgrade Your Browser" %}</h4
 
             check_popover();
 
-            var OpenMapSurfer_Roads = L.tileLayer('http://korona.geog.uni-heidelberg.de/tiles/roads/x={x}&y={y}&z={z}', {
-                zIndex: 0,
-                noWrap: true,
-                attribution: 'Imagery from <a href="http://giscience.uni-hd.de/">GIScience Research Group @ University of Heidelberg</a> &mdash; Map data &copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>'
-            });
             var osm = L.tileLayer('http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
```
Comment (lucernae) [inline templates/geosafe/analysis/create_base.html]:
```suggestion
            var default_basemap = L.tileLayer('http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
```

File: templates/geosafe/analysis/create_base.html [hunk 3]
```diff
@@ -526,21 +526,22 @@ <h4 class="modal-title" id="myModalLabel">{% trans "Upgrade Your Browser" %}</h4
...
+
+                    base_maps["{{ tile.0 }}"] = L.tileLayer('{{ tile.1|safe }}', {
+                        attribution: '{{ tile.2|safe }}'
+                });
+                {% endfor %}
+            {% endif %}
+
             map = L.map('map', {
-                layers: [OpenMapSurfer_Roads],
+                layers: [osm],
```
Comment (lucernae) [inline templates/geosafe/analysis/create_base.html]:
```suggestion
                layers: [default_basemap],
```

File: templates/geosafe/analysis/detail_base.html [hunk 4]
```diff
@@ -226,11 +226,7 @@ <h3 class="panel-title">{% trans "Action" %}</h3>
                 subdomains: '1234'
             });
 
-            var OpenMapSurfer_Roads = L.tileLayer('http://korona.geog.uni-heidelberg.de/tiles/roads/x={x}&y={y}&z={z}', {
-                noWrap: true,
-                attribution: 'Imagery from <a href="http://giscience.uni-hd.de/">GIScience Research Group @ University of Heidelberg</a> &mdash; Map data &copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>'
-            });
-            map.addLayer(OpenMapSurfer_Roads);
+            map.addLayer(osm);
```
Comment (lucernae) [inline templates/geosafe/analysis/detail_base.html]:
```suggestion
            map.addLayer(default_basemap);
```

**Review by NyakudyaA — COMMENTED** [review]
Nice @boney-bun , Does it also cover the layer details page where the default will be OSM

**Review by lucernae — APPROVED** [review]
LGTM @boney-bun , thank you :).

Could you check my suggestion too in your local environment first.

The aim is to have the default basemap automatically taken from the first TILES config if defined, if not, use default osm.

File: templates/geosafe/analysis/create_base.html (hunk 1 above)
Comment (boney-bun) [reply templates/geosafe/analysis/create_base.html:542]: changes it into:
`default_basemap = base_maps["{{ LEAFLET_CONFIG.TILES.0.0 }}"];`

**Comment by boney-bun** [comment]
thanks for the feedback @lucernae @NyakudyaA 
i've revised and tested it locally.
i'm going to merge this PR.

**Pull request merged**